### ISSUE TITLE
The MIME type is missing a parameter value for a parameter with the name 'charset'

### ISSUE THREAD

**Issue by mikenorgate** [OData/odata.net#165]
When running in Mono I get the following exception which I do not receive when running on Windows

> Microsoft.OData.Core.ODataContentTypeException: The MIME type is missing a parameter value for a parameter with the name 'charset'. at Microsoft.OData.Core.HttpUtils.ReadMediaTypeParameter (System.String text, System.Int32& textIndex, System.Collections.Generic.List1& parameters, System.String& charset) <0x40abf040 + 0x0039b> in <filename unknown>:0 
>     at Microsoft.OData.Core.HttpUtils.ReadMediaTypes (System.String text) <0x40abe7d0 + 0x0013f> in <filename unknown>:0 
>     at Microsoft.OData.Core.HttpUtils.MediaTypesFromString (System.String text) <0x40abe760 + 0x00047> in <filename unknown>:0 
>     at Microsoft.OData.Core.MediaTypeUtils.ParseContentType (System.String contentTypeHeader, System.String& charset) <0x40abe5e0 + 0x0003b> in <filename unknown>:0 
>     at Microsoft.OData.Core.MediaTypeUtils.GetFormatFromContentType (System.String contentTypeName, Microsoft.OData.Core.ODataPayloadKind[] supportedPayloadKinds, Microsoft.OData.Core.ODataMediaTypeResolver mediaTypeResolver, Microsoft.OData.Core.ODataMediaType& mediaType, System.Text.Encoding& encoding, Microsoft.OData.Core.ODataPayloadKind& selectedPayloadKind) <0x40abe000 + 0x000af> in <filename unknown>:0 
>     at Microsoft.OData.Core.MediaTypeUtils.GetFormatFromContentType (System.String contentTypeHeader, Microsoft.OData.Core.ODataPayloadKind[] supportedPayloadKinds, Microsoft.OData.Core.ODataMediaTypeResolver mediaTypeResolver, Microsoft.OData.Core.ODataMediaType& mediaType, System.Text.Encoding& encoding, Microsoft.OData.Core.ODataPayloadKind& selectedPayloadKind, System.String& batchBoundary) <0x40abdc90 + 0x0006b> in <filename unknown>:0 
>     at Microsoft.OData.Core.ODataMessageWriter.EnsureODataFormatAndContentType () <0x40ab91a0 + 0x00143> in <filename unknown>:0 
>     at Microsoft.OData.Core.ODataMessageWriter.SetHeaders (ODataPayloadKind payloadKind) <0x40ab8e40 + 0x00027> in <filename unknown>:0 
>     at Microsoft.OData.Core.ODataMessageWriter.SetOrVerifyHeaders (ODataPayloadKind payloadKind) <0x40ab8cf0 + 0x0002f> in <filename unknown>:0 
>     at Microsoft.OData.Core.ODataMessageWriter.WriteToOutput[TResult](ODataPayloadKind payloadKind, System.Action verifyHeaders, System.Func2 writeFunc) <0x40ab8a20 + 0x0002f> in :0 at Microsoft.OData.Core.ODataMessageWriter.CreateODataFeedWriter (IEdmEntitySetBase entitySet, IEdmEntityType entityType) <0x40ab8740 + 0x00173> in :0 at System.Web.OData.Formatter.Serialization.ODataFeedSerializer.WriteObject (System.Object graph, System.Type type, Microsoft.OData.Core.ODataMessageWriter messageWriter, System.Web.OData.Formatter.Serialization.ODataSerializerContext writeContext) <0x40ab7880 + 0x001a7> in :0 at System.Web.OData.Formatter.ODataMediaTypeFormatter.WriteToStream (System.Type type, System.Object value, System.IO.Stream writeStream, System.Net.Http.HttpContent content, System.Net.Http.Headers.HttpContentHeaders contentHeaders) <0x40ab0a90 + 0x009ed> in :0 at System.Web.OData.Formatter.ODataMediaTypeFormatter.WriteToStreamAsync (System.Type type, System.Object value, System.IO.Stream writeStream, System.Net.Http.HttpContent content, System.Net.TransportContext transportContext, CancellationToken cancellationToken) <0x40ab0870 + 0x000f7> in :0
## Enviroment

Ubuntu 14.10
Mono 4.1.0
.Net 4.5
## Request Headers

Accept:_/_
Accept-Encoding:gzip, deflate, sdch
Accept-Language:en-GB,en;q=0.8,pl;q=0.6,en-US;q=0.4
Cache-Control:no-cache
Connection:keep-alive
Content-Type:application/json;charset=utf-8
Host:192.168.0.2:5000
User-Agent:Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.118 Safari/537.36

**Comment by cmjacobs** [comment]
Has there been any further news on this issue?

**Comment by cmjacobs** [comment]
I think I figured this one out. It’s always nice to be able to see the source code.

**tl;dr** The request must include HTTP header “Accept-Charset=utf-8”.

Full explanation:
This is not actually a bug in the OData .Net Libraries. This is a side effect of a (flawed) difference between the Windows and Mono implementation of System.Net.Http. The place where this could be fixed is in System.Web.OData.

In System.Web.OData, there is a method in ODataMediaTypeFormatter called SetDefaultContentHeaders. In there, there is a check to see if the charset of content is in the list of accepted charset’s requested by the client. If not, the charset attribute in the header is “cleaned.”
```cs
// In general, in Web API we pick a default charset based on the supported character sets
// of the formatter. However, according to the OData spec, the service shouldn't be sending
// a character set unless explicitly specified, so if the client didn't send the charset we chose
// we just clean it.
if (headers.ContentType != null &&
    !Request.Headers.AcceptCharset
            .Any(cs => cs.Value.Equals(headers.ContentType.CharSet, StringComparison.OrdinalIgnoreCase)))
{
    headers.ContentType.CharSet = String.Empty;
}
```
In the Windows implementation of MediaTypeHeaderValue (in System.Net.Http.dll, namespace: System.Net.Http.Headers), when CharSet is set to String.Empty, charset is removed from parameters collection. In the mono implementation, the empty value is left in the collection. Specifically, the extension method SetValue in CollectionExtensions.cs does not account for empty strings.

When the response is being formatted, the headers are processed and key-value pairs are expected. Because the mono implementation leaves an empty value for charset, the above exception is thrown.

There are a couple of workarounds. The easiest: you can have the client send Accept-Charset=utf-8 with the OData requests. 

Since I am using OWIN, I inserted a small piece of middleware that injects the Accept-Charset header if it does not exist.
```cs
public class InjectAcceptCharsetMiddleware : OwinMiddleware
{
    public InjectAcceptCharsetMiddleware(OwinMiddleware next, IAppBuilder builder) : base(next) { }
 
    public override async Task Invoke(IOwinContext context)
    {
 
        if (!context.Request.Headers.ContainsKey("Accept-Charset"))
            context.Request.Headers.Add("Accept-Charset", new[] { "utf-8" });
 
        await Next.Invoke(context);
 
    }
}
```

I hope this helps.
